### PR TITLE
(CDAP-16128) Added batch get app details endpoint

### PR DESCRIPTION
**Pull request by chtyim**
Added a new endpoint /appdetail for batch fetching application details
for a given set of application name+version in a given namespace.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java [hunk 1]
```diff
@@ -364,6 +372,77 @@ public void updateApp(FullHttpRequest request, HttpResponder responder,
...
+   * }
+   * </pre>
+   *
+   * The response will be an array of {@link BatchApplicationDetail} object, which either indicates a success (200) or
+   * failure for each of the requested application in the same order as the request.
+   */
+  @POST
+  @Path("/appdetail")
+  public void getApplicationDetails(FullHttpRequest request, HttpResponder responder,
+                                    @PathParam("namespace-id") String namespace) throws Exception {
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java:395]: should be final

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java [hunk 2]
```diff
@@ -364,6 +372,77 @@ public void updateApp(FullHttpRequest request, HttpResponder responder,
...
+   *   {"appId":"FOO", "version":"2.3.4"},
+   * ]
+   * }
+   * </pre>
+   *
+   * The response will be an array of {@link BatchApplicationDetail} object, which either indicates a success (200) or
+   * failure for each of the requested application in the same order as the request.
+   */
+  @POST
+  @Path("/appdetail")
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java:393]: what do you think about calling it `appsdetail`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java [hunk 3]
```diff
@@ -364,6 +372,77 @@ public void updateApp(FullHttpRequest request, HttpResponder responder,
...
+      if (array == null) {
+        throw new BadRequestException("Request body is invalid json, please check that it is a json array.");
+      }
+      for (JsonElement element : array) {
+        if (!element.isJsonObject()) {
+          throw new BadRequestException("Request element is expected to be a json object.");
+        }
+        JsonObject obj = element.getAsJsonObject();
+        if (!obj.has("appId")) {
+          throw new BadRequestException("Missing 'appId' in the request element");
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: nit: period at the end of error message

**Review by CuriousVini — APPROVED** [review]
minor comments. Code change lgtm 👍

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java (hunk 1 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java:395]: Generally we don't declare local variables as `final`.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java (hunk 2 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java:393]: I was following the existing convention for batch endpoints that they don't use plural.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java (hunk 3 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: fixed

**Comment by chtyim** [comment]
Build passed https://builds.cask.co/browse/CDAP-RUT1729

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java (hunk 1 above)
Comment (CuriousVini) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java:395]: I noticed it at multiple places in the same class.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java (hunk 2 above)
Comment (CuriousVini) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java:393]: for `/runs`, `/apps` endpoints we use plural.

Comment (chtyim) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java:393]: We do have some insistency for batch endpoint. Some are singular `/status`, `/runcount`, `/start`, `/stop`, while only the `runs` is plural. For `/apps` and `/runs`, they are representing collections.

I am more inclined to use singular for batch endpoint that does something different than the collection. It is acted more like a command.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java (hunk 1 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java:395]: Those are old code on Java 7 that needs an explicit `final` to be used in inner class. However, when we move to Java 8 and also when moving out some of the inner classes usages, we haven't remove all those `final` identifier.

**Pull request merged**